src/cplscheme/MultiCouplingScheme.cpp

```diff
@@ -122,73 +88,58 @@ void MultiCouplingScheme::initializeData()
     }
     sendData();
   }
-
-  // in order to check in advance if initializeData has been called (if necessary)
-  setHasToSendInitData(false);
-  setHasToReceiveInitData(false);
 }
 
-void MultiCouplingScheme::advance()
+void MultiCouplingScheme::explicitAdvance()
 {
-  PRECICE_TRACE(getTimeWindows(), getTime());
-  checkCompletenessRequiredActions();
-
-  PRECICE_CHECK(!hasToReceiveInitData() && !hasToSendInitData(),
-                "initializeData() needs to be called before advance if data has to be initialized!");
+  PRECICE_ASSERT(isImplicitCouplingScheme(), "MultiCouplingScheme is always Implicit.");
+  // TODO this class hierarchy has a smell! I think that MultiCouplingScheme should not be derived from BaseCouplingScheme, but directly from CouplingScheme or an intermediate layer between CouplingScheme and BaseCouplingScheme.
+}
 
-  setHasDataBeenExchanged(false);
-  setIsTimeWindowComplete(false);
-  bool convergence = false;
-  if (math::equals(getThisTimeWindowRemainder(), 0.0, _eps)) {
-    PRECICE_DEBUG("Computed full length of iteration");
+std::pair<bool, bool> MultiCouplingScheme::implicitAdvance()
+{
+  PRECICE_ASSERT(isImplicitCouplingScheme(), "MultiCouplingScheme is always Implicit.");
+  PRECICE_DEBUG("Computed full length of iteration");
 
-    receiveData();
+  receiveData();
 
-    auto designSpecifications = getAcceleration()->getDesignSpecification(_allData);
-    convergence               = measureConvergence(designSpecifications);
+  auto designSpecifications = getAcceleration()->getDesignSpecification(_allData);
+  bool convergence          = measureConvergence(designSpecifications);
 
-    // Stop, when maximal iteration count (given in config) is reached
-    if (maxIterationsReached()) {
-      convergence = true;
-    }
-    if (convergence) {
-      if (getAcceleration().get() != nullptr) {
-        getAcceleration()->iterationsConverged(_allData);
-      }
-      newConvergenceMeasurements();
-      timeWindowCompleted();
-    } else if (getAcceleration().get() != nullptr) {
-      getAcceleration()->performAcceleration(_allData);
+  // Stop, when maximal iteration count (given in config) is reached
+  if (maxIterationsReached()) {
+    convergence = true;
+  }
+  if (convergence) {
+    if (getAcceleration().get() != nullptr) {
+      getAcceleration()->iterationsConverged(_allData);
     }
+    newConvergenceMeasurements();
+    timeWindowCompleted();
+  } else if (getAcceleration().get() != nullptr) {
+    getAcceleration()->performAcceleration(_allData);
```

Comment:
I have the feeling that all these calls could also go to `BaseCouplingScheme`. The only difference between the coupling schemes is on which data the acceleration is used on.

Reply:
So this means that we want an additional (common!) building block for acceleration inside `BaseCouplingScheme::advance()`:

```
  if (subcyclingIsCompleted()) {
    if (isExplicitCouplingScheme()) {
      timeWindowCompleted();
      explicitAdvance();
      _computedTimeWindowPart = 0.0;
    } else if (isImplicitCouplingScheme()) {
      std::pair<bool, bool> convergenceInformation = implicitAdvance();
      // accelerate building-block start >>>
      if (getAcceleration().get() != nullptr) {
        getAcceleration()->performAcceleration(_allData);
      }
      // accelerate building-block end <<<
      bool convergence = convergenceInformation.first;
      bool convergenceCoarseOptimization = convergenceInformation.second;
      if (not convergence) {
        PRECICE_DEBUG("No convergence achieved");
        requireAction(constants::actionReadIterationCheckpoint());
      } else {
        PRECICE_DEBUG("Convergence achieved");
        advanceTXTWriters();
      }
      updateTimeAndIterations(convergence, convergenceCoarseOptimization);
    }
  }
```

Possibly this also allows us to merge `explicitAdvance()` and `implicitAdvance()` into a specific implementation of `exchangeData()`, depending on which `BaseCouplingScheme` is used.

Reply:
For easier implementation, I will only consider `SerialCouplingScheme` and `ParallelCouplingScheme` for these steps for now. This means I will drop the `final` in BaseCouplingScheme::advance()` to let `MultiCouplingScheme` override its own (temporary) `advance`. We can fix the `MultiCouplingScheme`, as soon as the other two are in good shape.

Reply:
dde4e2f1fbfe64b6462831c461414412da48e784 and 8093c8daefeabfbc3eb632ee33cb4e5b744627aa provide some first steps into this direction.

Reply:
I added two new functions in `BaseCouplingScheme` in 3d2e44a. This considerably shortens `implicitAdvance` in the derived classes `SerialCouplingScheme` and `ParallelCouplingScheme`. `MultiCouplingScheme` needs further consideration, since it currently does not fit into this pattern, at all.

@uekerman some questions popped up during refactoring, I marked them in the code. I would suggest to consider this conversation as resolved (it is a good intermediate state and the acceleration is factored out) and start another reviewing loop, where we try to improve furtherly as soon as the other todos from your first review are done.

Reply:
See below. https://github.com/precice/precice/pull/689#issuecomment-596738688

Reply:
I implemented your suggestions in 151b263. There are still some issues, but I think it is sufficient to show the consequences of restructuring as suggested. @uekerman What do you think?